src/provider/repository/stackablerepository.rs

```diff
@@ -0,0 +1,197 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::fmt;
+use std::fs::File;
+use std::hash::{Hash, Hasher};
+use std::io::{copy, Cursor};
+use std::path::PathBuf;
+
+use kube::api::Meta;
+use log::{debug, trace};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+use crate::provider::error::StackableError;
+use crate::provider::error::StackableError::PackageNotFound;
+use crate::provider::repository::package::Package;
+use crate::provider::repository::repository::Repository;
+
+#[derive(Debug, Clone)]
+pub struct StackableRepoProvider {
+    base_url: Url,
+    pub name: String,
+    content: Option<RepositoryContent>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct RepoData {
+    version: String,
+    parcels: HashMap<String, Vec<Product>>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Product {
+    version: String,
+    path: String,
+    hashes: HashMap<String, String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct RepositoryContent {
+    pub version: String,
+    pub parcels: HashMap<String, HashMap<String, StackablePackage>>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct StackablePackage {
+    pub product: String,
+    pub version: String,
+    pub link: String,
+    pub hashes: HashMap<String, String>,
+}
+
+impl StackableRepoProvider {
+    pub fn new(name: String, base_url: String) -> Result<StackableRepoProvider, StackableError> {
+        let base_url = Url::parse(&base_url)?;
+
+        Ok(StackableRepoProvider { base_url, name, content: None })
+    }
+
+    pub async fn provides_package<T: Into<Package>>(&mut self, package: T) -> Result<bool, StackableError> {
+        debug!("Starting metadata refresh for repository of type {} at location {}", "StackableRepo", self.base_url);
+        let package = package.into();
+        let metadata = self.get_repo_metadata().await?;
+        debug!("Repository provides the following products: {:?}", metadata);
+        if let Some(product) = metadata.parcels.get(&package.product) {
+            return Ok(product.contains_key(&package.version));
+        }
+        Ok(false)
+    }
+
+    async fn get_package(&mut self, package: Package) -> Result<StackablePackage, StackableError> {
```

Comment:
I'm wondering if we can just take a reference here but you probably want an owned object for the error message.

How can I clone an object via its reference? I have no idea...